File: src/test/java/com/josdem/vetlog/client/GoogleStorageWriterTest.java
package com.josdem.vetlog.client;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.google.api.gax.core.CredentialsProvider;
import com.google.auth.Credentials;
import com.google.cloud.spring.core.GcpProjectIdProvider;
import com.google.cloud.storage.Storage;
import com.google.cloud.storage.StorageOptions;
import com.josdem.vetlog.helper.StorageOptionsHelper;
import java.io.InputStream;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInfo;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

@Slf4j
class GoogleStorageWriterTest {

    private GoogleStorageWriter googleStorageWriter;

    @Mock
    private CredentialsProvider credentialsProvider;

    @Mock
    private GcpProjectIdProvider gcpProjectIdProvider;

    @Mock
    private StorageOptionsHelper storageOptionsHelper;

    @BeforeEach
    void setup() {
        MockitoAnnotations.openMocks(this);
        googleStorageWriter = new GoogleStorageWriter(credentialsProvider, gcpProjectIdProvider, storageOptionsHelper);
    }

    @Test
    @DisplayName("Should upload to bucket")
    void shouldUploadToBucket(TestInfo testInfo) throws Exception {
        log.info(testInfo.getDisplayName());
        var inputStream = mock(InputStream.class);
        var builder = mock(StorageOptions.Builder.class);
        var credentials = mock(Credentials.class);
        var storage = mock(Storage.class);
        var storageOptions = mock(StorageOptions.class);

        when(gcpProjectIdProvider.getProjectId()).thenReturn("projectId");
        when(credentialsProvider.getCredentials()).thenReturn(credentials);

        when(storageOptionsHelper.getStorageOptions()).thenReturn(builder);
        when(builder.setProjectId("projectId")).thenReturn(builder);
        when(builder.setCredentials(credentials)).thenReturn(builder);
        when(builder.build()).thenReturn(storageOptions);
        when(builder.build().getService()).thenReturn(storage);

        googleStorageWriter.setup();
        googleStorageWriter.uploadToBucket("bucket", "fileName", inputStream, "contentType");
        verify(inputStream).readAllBytes();
    }
}

File: src/test/java/com/josdem/vetlog/client/GoogleStorageWriterTest.kt
package com.josdem.vetlog.client

import com.google.api.gax.core.CredentialsProvider
import com.google.auth.Credentials
import com.google.cloud.spring.core.GcpProjectIdProvider
import com.google.cloud.storage.Storage
import com.google.cloud.storage.StorageOptions
import com.josdem.vetlog.exception.BusinessException
import com.josdem.vetlog.helper.StorageOptionsHelper
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInfo
import org.junit.jupiter.api.assertThrows
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.mockito.MockitoAnnotations
import org.slf4j.LoggerFactory
import java.io.InputStream

class GoogleStorageWriterTest {
    private lateinit var googleStorageWriter: GoogleStorageWriter

    @Mock
    private lateinit var credentialsProvider: CredentialsProvider

    @Mock
    private lateinit var gcpProjectIdProvider: GcpProjectIdProvider

    @Mock
    private lateinit var storageOptionsHelper: StorageOptionsHelper

    @Mock
    private lateinit var inputStream: InputStream

    @Mock
    private lateinit var storage: Storage

    @Mock
    private lateinit var credentials: Credentials

    @Mock
    private lateinit var storageOptions: StorageOptions

    @Mock
    private lateinit var builder: StorageOptions.Builder

    companion object {
        private val log = LoggerFactory.getLogger(this::class.java)
    }

    @BeforeEach
    fun setup() {
        MockitoAnnotations.openMocks(this)
        googleStorageWriter = GoogleStorageWriter(credentialsProvider, gcpProjectIdProvider, storageOptionsHelper)
    }

    @Test
    fun `should upload to bucket`(testInfo: TestInfo) {
        log.info(testInfo.displayName)
        setExpectations()

        googleStorageWriter.uploadToBucket("bucket", "fileName", inputStream, "contentType")
        verify(inputStream).readAllBytes()
    }

    @Test
    fun `should not upload to bucket`(testInfo: TestInfo) {
        log.info(testInfo.displayName)
        setExpectations()

        `when`(inputStream.readAllBytes()).thenThrow(IllegalStateException("Error"))
        assertThrows<BusinessException> {
            googleStorageWriter.uploadToBucket("bucket", "fileName", inputStream, "contentType")
        }
    }

    private fun setExpectations() {
        `when`(gcpProjectIdProvider.projectId).thenReturn("projectId")
        `when`(credentialsProvider.credentials).thenReturn(credentials)

        `when`(storageOptionsHelper.storageOptions).thenReturn(builder)
        `when`(builder.setProjectId("projectId")).thenReturn(builder)
        `when`(builder.setCredentials(credentials)).thenReturn(builder)
        `when`(builder.build()).thenReturn(storageOptions)
        `when`(builder.build().service).thenReturn(storage)
    }
}
